feat(compose): Support depends_on property

LucaSeri opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

  Kraftkit follows the same guidelines as the Unikraft Open Source Project.

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

  - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
  - [x] Tested your changes against relevant architectures and platforms;
  - [x] Ran `make fmt` on your commit series before opening this PR;
  - [x] Updated relevant documentation.

### Description of changes
The depends_on property allows specifying dependencies between services. The services will always be started in the order of the dependencies and always be stopped in the opposite order. At the moment we cannot do healthchecks in krafkit, so the condition of the dependency is ignored.

E.g.
```yaml
services:
  a:   
    image: nginx:1.15
    depends_on:
      - b
  b:
    image: nginx:1.15
```

``kraft compose up -d``

```
 i  creating network test_default...
 i  creating service b...
 i  creating service a..
```

```kraft compose down```

```
 i  removing service a...
 i  removing service b...
 i  removing network test_default..
```
<!--
Please provide a detailed description of the changes made in this new PR.
-->

Commit:
feat(compose): Support depends_on property

The depends_on property allows specifying dependencies
between services. The services will always be started
in the order of the dependencies and always be stopped
in the opposite order. At the moment we cannot do healthchecks
in krafkit, so the condition of the dependency is ignored.

Signed-off-by: Luca Seritan <luca.seritan@gmail.com>

nderjung commented on internal/cli/kraft/compose/utils/utils.go:
```diff
@@ -73,3 +74,77 @@ func RemoveOrphans(ctx context.Context, project *compose.Project) error {
+// ServicesReversedWithDependants receives a list of services and generates a
+// new list reverse ordered by dependencies. If expand is set, it will also
+// include dependendants not present in the original list.
+func ServicesReversedByDependencies(ctx context.Context, project *compose.Project, services types.Services, expand bool) []types.ServiceConfig {
```
Can you place this inside of the `compose` package?

nderjung commented on internal/cli/kraft/compose/utils/utils.go:
```diff
@@ -73,3 +74,77 @@ func RemoveOrphans(ctx context.Context, project *compose.Project) error {
+				continue
+			}
+
+			log.G(ctx).Debugf("adding dependant %s -> %s", service.Name, name)
```
```go
log.G(ctx).
  WithField("service", service.Name).
  WithField("on", name).
  Debug("depends")
```

nderjung commented on internal/cli/kraft/compose/utils/utils.go:
```diff
@@ -73,3 +74,77 @@ func RemoveOrphans(ctx context.Context, project *compose.Project) error {
+				continue
+			}
+
+			log.G(ctx).Debugf("adding dependency %s -> %s", service.Name, name)
```
```go
log.G(ctx).
  WithField("service", service.Name).
  WithField("on", name).
  Debug("depends")
```

nderjung commented on internal/cli/kraft/compose/utils/utils.go:
```diff
@@ -73,3 +74,77 @@ func RemoveOrphans(ctx context.Context, project *compose.Project) error {
+// ServicesOrderedByDependencies receives a list of services and generates a
+// new list ordered by dependencies. If expand is set, it will also include
+// dependencies not present in the original list.
+func ServicesOrderedByDependencies(ctx context.Context, project *compose.Project, services types.Services, expand bool) []types.ServiceConfig {
```
Can you move this method into the `compose` package?

nderjung commented:
Nice! Can you split this into two commits, one which introduces `ServicesOrderedByDependencies` and `ServicesOrderedByDependencies` into the `compose` package and another which utilizes these new methods across the `internal/cli/kraft/compose` subcommands? Thanks!

Commit:
feat(compose): Introduce methods for ordering services by dependencies

The two newly introduced methods help with fetching dependencies
and ordering services either by dependency order or reverse
dependency order.

Signed-off-by: Luca Seritan <luca.seritan@gmail.com>

nderjung reviewed (approved):
Awesome, thanks!
